html/semantics/forms/form-submission-0/historical.window.js

```diff
@@ -0,0 +1,18 @@
+// META: script=./resources/targetted-form.js
+
+test(t => {
+  const form = populateForm('<input required><input type=submit>');
+  t.add_cleanup(() => {
+    form.previousElementSibling.remove();
+    form.remove();
+  });
+  const submitter = form.querySelector('input[type=submit]');
+  let invalid = form.querySelector('[required]');
+  let counter = 0;
+  form.addEventListener("invalid", t.step_func(() => counter++));
+  form.oninvalid = t.step_func(() => counter++);
+  invalid.addEventListener("invalid", t.step_func(() => counter++));
+  invalid.oninvalid = t.step_func(() => counter++);
+  submitter.click();
+  assert_equals(counter, 2);
```

Comment:
A browser could theoretically pass this if it fired invalid at both the form and the control, but did not support the oninvalid property on either.

I think this would be better if each function pushed a unique value into an array, and then you did `assert_array_equals`. Or at least two separate counters, one for the form and one for the control.

Reply:
Fair, pushed something better.